### PR TITLE
Fix the derivative of `MottonenStatePreparation` where possible

### PR DESCRIPTION
**Pull request by dwierichs**
**Context:**
The decomposition of `MottonenStatePreparation` skips some gates for special parameter values/input states. 
See the linked issue for details.

**Description of the Change:**
This PR introduces a check for differentiability so that the gates only are skipped when no derivatives are being computed.
Note that this does *not* fix the non-differentiability at other special parameter points that also is referenced in #5715 and that is being warned against in the docs already.
Also, the linked issue is about multiple operations and we here only address `MottonenStatePreparation`.

**Benefits:**
Fixes parts of #5715. Unblocks #5620 .

**Possible Drawbacks:**

**Related GitHub Issues:**
#5715

### CODE REVIEW

File: tests/templates/test_state_preparations/test_mottonen_state_prep.py [hunk 1]
```diff
@@ -67,74 +67,60 @@ def test_get_alpha_y(self, current_qubit, expected, tol):
...
+    (x := np.array([2, 0, 0, 0, 1, 0, 0, 2]) / 3, [0, 1, 2], x),
+    (x := np.array([1, 1j, 1, -1j, 1, 1, 1, 1j]) / np.sqrt(8), [0, 1, 2], x),
+    (fixed_states[0], [0, 1, 2], fixed_states[0]),
+    (fixed_states[1], [0, 1, 2], fixed_states[1]),
+    (fixed_states[2], [0, 1, 2], fixed_states[2]),
+    (fixed_states[3], [0, 1, 2], fixed_states[3]),
+    (x := np.array([1 / 2, 0, 0, 0, 1j / 2, 0, 1j / np.sqrt(2), 0]), [0, 1, 2], x),
+    (np.array([1 / 2, 0, 1j / 2, 1j / np.sqrt(2)]), [0, 1], x),
+]
+
```
Comment (dwierichs) [inline tests/templates/test_state_preparations/test_mottonen_state_prep.py:118]: This is a mere reformatting step.

File: tests/templates/test_state_preparations/test_mottonen_state_prep.py [hunk 2]
```diff
@@ -492,3 +414,32 @@ def circuit(state):
...
+    params = jax.numpy.array([0.5, 0.5, 0.5, 0.5])
+    jac_exact_fn = jax.jacobian(circuit_exact)
+    jac_fn = jax.jacobian(circuit_fd)
+    jac_jit_fn = jax.jit(jac_fn)
+
+    jac_exact = jac_exact_fn(params)
+    jac = jac_fn(params)
+    jac_jit = jac_jit_fn(params)
+
+    assert qml.math.allclose(jac_exact, jac_jit, atol=0.1)
```
Comment (astralcai) [inline tests/templates/test_state_preparations/test_mottonen_state_prep.py:444]: `atol=0.1` seems a bit too large, I believe the standard tolerance for stochastic results is typically `0.05`

File: tests/templates/test_subroutines/test_qubitization.py [hunk 3]
```diff
@@ -255,10 +257,9 @@ def test_qnode_jax(self, shots, use_jit, device):
         assert jac.shape == (4,)
         assert np.allclose(jac, self.exp_grad, atol=0.05)
 
+    # TODO: finite shots fails because Prod is not currently differentiable.
     @pytest.mark.torch
-    @pytest.mark.parametrize(
-        "shots", [None]
-    )  # TODO: finite shots fails because Prod is not currently differentiable.
+    @pytest.mark.parametrize("shots", [None])
```
Comment (astralcai) [inline tests/templates/test_subroutines/test_qubitization.py:262]: Maybe add finite shots to the parametrization but `xfail` inside the test if `shots is not None`

File: tests/templates/test_subroutines/test_qubitization.py [hunk 4]
```diff
@@ -222,11 +222,13 @@ def test_qnode_autograd(self):
         assert qml.math.shape(res) == (4,)
         assert np.allclose(res, self.exp_grad, atol=1e-5)
 
+    # TODO: Allow the following cases once their underlying issues are fixed:
+    #  (True, 50000): jax.jit on jax.grad does not work with AmplitudeEmbedding currently
+    #  (False, 50000): Since #5774, the decomposition of AmplitudeEmbedding triggered by
+    #                  param-shift includes a GlobalPhase always. GlobalPhase will only be
+    #                  param-shift-compatible again once #5620 is merged in.
```
Comment (astralcai) [inline tests/templates/test_subroutines/test_qubitization.py:229]: Maybe include them in the parametrization but `xfail` these special cases inside the test case.

**Review by astralcai — COMMENTED** [review]
Just a small suggestion regarding skipping test cases, otherwise LGTM!

**Review by astralcai — APPROVED** [review]
LGTM! 🎸

File: tests/templates/test_state_preparations/test_mottonen_state_prep.py [hunk 5]
```diff
@@ -492,3 +414,32 @@ def circuit(state):
     expected = np.zeros(8)
     expected[0] = 1.0
     assert qml.math.allclose(actual, expected)
+
+
+@pytest.mark.jax
+@pytest.mark.parametrize("shots, atol", [(None, 0.005), (1000000, 0.05)])
```
[the remark this answers is not in the document]
Comment (dwierichs) [reply tests/templates/test_state_preparations/test_mottonen_state_prep.py:420]: Regarding the output format: The derivative of a real-valued function with respect to complex-valued input parameters is complex-valued in general.

**Review by dwierichs — COMMENTED** [review]
I answered both the complex numbers and precision issues in the comment :)

**Review by KetpuntoG — APPROVED** [review]
Thanks for the clarifications! Ready to go 🚀

**Pull request merged**